tool/tsh/common/db.go

```diff
@@ -1248,115 +1292,95 @@ func isMFADatabaseAccessRequired(ctx context.Context, tc *client.TeleportClient,
 //
 // If logged into multiple databases, returns an error unless one specified
 // explicitly via --db flag.
-func pickActiveDatabase(cf *CLIConf, tc *client.TeleportClient) (*tlsca.RouteToDatabase, error) {
-	profile, err := tc.ProfileStatus()
-	if err != nil {
-		return nil, trace.Wrap(err)
+func pickActiveDatabase(cf *CLIConf, tc *client.TeleportClient, routes []tlsca.RouteToDatabase) (*tlsca.RouteToDatabase, error) {
+	selectors := newDatabaseResourceSelectors(cf)
+	// first check if an active route can even potentially match.
+	routes = filterRoutesByPrefix(routes, selectors.name)
+	if len(routes) == 0 {
+		return nil, trace.NotFound(formatDbNotLoggedIn(cf.SiteName, selectors))
 	}
-
-	routes, err := profile.DatabasesForCluster(tc.SiteName)
+	info, err := getDatabaseInfo(cf, tc, routes)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
+	if !info.isActive {
+		return nil, trace.NotFound(formatDbNotLoggedIn(cf.SiteName, selectors))
+	}
+	return &info.RouteToDatabase, nil
+}
 
-	if len(routes) == 0 {
-		return nil, trace.NotFound("please login using 'tsh db login' first")
+// getDatabasesForLogout takes a list of active database routes and returns a
+// selected list of databases to logout.
+// Logout selection is done by matching on database name/prefix, label, query
+// predicate selectors from the Teleport client.
+func getDatabasesForLogout(cf *CLIConf, tc *client.TeleportClient, activeRoutes []tlsca.RouteToDatabase) ([]tlsca.RouteToDatabase, error) {
+	selectors := newDatabaseResourceSelectors(cf)
+	activeRoutes = filterRoutesByPrefix(activeRoutes, selectors.name)
+	if len(activeRoutes) == 0 || selectors.IsEmpty() {
+		// nothing to filter further
+		return activeRoutes, nil
+	}
+	if route, ok := maybeFindActiveDatabase(selectors, activeRoutes); ok {
```

Comment:
How is the rest of this function starting here different from `pickActiveDatabase`? To me this is a better implementation. Could we merge the two?

Reply:
In a couple of ways.  This function can return multiple databases for logout. It can also select an active route to logout of even when the database is no longer registered.

Whereas `pickActiveDatabase` needs to get exactly one database, and it should ignore active routes that correspond to databases that no longer exist.

Reply:
fwiw, this function was originally just `filterActiveDatabases`, and it was used by both `pickActiveDatabase` and `onDatabaseLogout`. I find it easier to understand the code if I keep `tsh db logic` selection logic separate from the other `tsh db` commands, because it's uniquely able to select multiple databases unlike the other commands.

So this way, `tsh db logout` does its special case logic in this func, and all other `tsh db` commands flow through `getDatabaseInfo` to get a single database

Reply:
> This function can return multiple databases for logout. 

For example, say there are two databases `aurora` and `aurora-reader`
- `tsh db logout aurora` logs out one database
- `tsh db logout auror` or `tsh db logout aurora-` logs out both databases?

To be honest, personally, I would just ask the user to pick one to log out in the later cases.

>  It can also select an active route to logout of even when the database is no longer registered.

That's a good point.  Just now we have three version of very similar logic (connect/proxy, env/config, logout). The next person working on changing this logic will definitely have a hard time =p

Reply:
what changes would you like to see?

1. make `tsh db logout <some name/labels/query>` requires only one database selected? So it disallows logging out of multiple except in the case where they just do `tsh db logout` and it logs out of all?
2. config/env will just pick the active database regardless of whether it exists in backend? note that `tsh db ls` has never displayed logged in databases that aren't registered in the backend anymore. so it's kinda like a phantom cert in their profile if we don't ignore it.

Reply:
> Now, if they have a logged in db that doesn't exist, it can interfere with selecting a registered database if the name matches exactly.

Could you give an example of this case?

Maybe disallow prefix match when <db-service-name> is specified? just exact-full-name or exactly-one-discover-name?

Reply:
This is on master:

```
$ tsh db ls env=test
tsh dName                 Description                   Allowed Users Labels   Connect
-------------------- ----------------------------- ------------- -------- -------------------
> foo (user: alice)  Example dyn database resource [*]           env=test tsh db connect foo
> foo2 (user: alice) Example dyn database resource [*]           env=test tsh db connect foo2

$ tsh db env foo
export MYSQL_GROUP_SUFFIX=_gavin-leaf.cloud.gravitational.io-foo

$ tctl rm db/foo
database "foo" has been deleted

$ tsh db ls env=test
Name                 Description                   Allowed Users Labels   Connect
-------------------- ----------------------------- ------------- -------- -------------------
> foo2 (user: alice) Example dyn database resource [*]           env=test tsh db connect foo2

$ tsh db env foo
export MYSQL_GROUP_SUFFIX=_gavin-leaf.cloud.gravitational.io-foo

$ tsh db logout foo
Logged out of database foo

$ tsh db logout foo
Logged out of database foo2
```

Reply:
I figured this might screw up a script after the v13->v14 upgrade.

Suppose they're logged into `mysql-rds`.
Then discovery agent upgrades to v14 and it renames it to `mysql-rds-etc...`
then the script does `tsh db login mysql-rds && tsh db [env | config] mysql-rds`
The script will login to `mysql-rds-etc...`, then get config for the stale cert they had instead of the database they just logged into

Maybe I'm just overthinking it. It's not terrible lol.